### ISSUE TITLE
Test fixes for `LockSubsystemCommand`

### ISSUE THREAD

**Issue by Samasaur1** [Roobotics-FRC/RooBotPost2017#5]
Our current approach is which is one `LockSubsystemCommand` that is passed a `Subsystem`. The `LockSubsystemCommand` then assigns the `Subsystem` it is passed to a variable called `lockedSubsystem`, and requires `lockedSubsystem` it is passed. This seems like it should work, but (as of last testing) it deploys successfully but crashes when enabled.

When the `Subsystem` calls `setDefaultCommand(new LockSubsystemCommand)` inside `initDefaultCommand`, the important part is this:

      boolean found = false;
      Enumeration requirements = command.getRequirements();
      while (requirements.hasMoreElements()) {
        if (requirements.nextElement().equals(this)) {
          found = true;
        }
      }
      if (!found) {
        throw new IllegalUseOfCommandException("A default command must require the subsystem"); // This is the error we are getting
      }
      m_defaultCommand = command;

This says that if this subsystem is found in the command requirements, it can be the default command. Our command is throwing the error `"A default command must require the subsystem"`, which means it is not being found. As we are passing the instance of the subsystem to the `LockSubsystemCommand`, this shouldn't happen (see below)

When `requires(this.lockedSubsytem)` is called, it is added to a `m_requirements` set defined in the `Command` class.

You may be wondering how these are connected. Here is the `getRequirements()` command in the `Command` class:

    synchronized Enumeration getRequirements() {
        return m_requirements == null ? emptyEnumeration : m_requirements.getElements();
    }

***

The flow is this:
+ The subsystem (in this example, the DriveTrain) calls `setDefaultCommand(new LockSubsystemCommand(this));`
  + This actually calls the `LockSubsystemCommand` constructor first, which adds the DriveTrain to its `m_requirements` set
  + Then we finish the `setDefaultCommand` method, which checks if it is in the command's requirements set. **This is where we are getting our exception.**

Please note that *I do not believe that we tested the code using Alex's `getDriveTrain()` fix* (using the DriveTrain as an example). However, we *did* try `DriveTrain.getDriveTrain()` and it **did not work.**
***
In short, I'm really stuck here. I don't think Alex's solution will work (but it *will* compile), so here are the options I see:

**1. Ignore the problem.** This will work in the short term. We haven't been able to replicate this problem, but if it happens again it will be pretty bad.
**2. Write a locking command for every subsystem.** This is the best option I see. It will fix the problem, and will not crash the robot. However, it is not great practice
**3. Write a locking command for only critical subsystems.** This is, in my opinion, the worst option. If we are willing to write a locking command for one, we should write one for every one. As Alex said earlier, *this same problem could be hiding in any of our subsystems right now.*

I'm going to leave the code how it is, but it **will not run on the robot in this state**, and the code on the robot no will not work either. I'd go with the second option, but a) I might be missing something, b) I don't *really* know what I'm talking about; all I did was look through some of WPILib, and c) *I on't really think I have the authority to make this decision*

What do you guys think? @thefangbear @aaplmath

**Comment by thefangbear** [comment]
Have we tried the following?

1. `super()` then `require()`
2. `require()` in the initialize() method

**Comment by thefangbear** [comment]
Please figure out how `command.getRequirements()` work and how to properly add commands. Our problem lies not in the design pattern but in correctly adding requirements.

**Comment by thefangbear** [comment]
And for your other try @Samasaur1 did you modify all subsystems? Only changing it for DriveTrain will also not work.

**Comment by Samasaur1** [comment]
@thefangbear No, we did not try `super()` then `require()` or putting `require()` in `initialize()`. I did indeed modify all subsystems, `DriveTrain` was the example because it was the one we found the error in.

I'll see if I can find out more about `Command.getRequirements()`.

**Comment by thefangbear** [comment]
Great! That adds to our wishlist for next week then. In the meanwhile, it might help to explore the abstract class `Command`. Pay specific notice to the constructor and getter/setters.

**Comment by thefangbear** [comment]
Sorry! It was my careless mistake. This issue shall be fixed since https://github.com/Roobotics-FRC/RooBotPost2017/commit/2ff494355ce280fdf14dd899cb86d167df9b91bf. I'll leave this open without the 🐛  tag until it's tested next time.

**Comment by thefangbear** [comment]
Sadly applying this because we won't really have a chance after the season's started and a change of priority.
